Validation errors and abort under swiftshader

When running `matrices_uniform_draw.amber`

```
[ERROR] validation layer (Validation):
 [ VUID-vkFreeMemory-memory-00677 ] Object: 0x250000000025 (Type = 8) | Cannot call vkFreeMemory on VkDeviceMemory 0x250000000025[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to memory (via images or buffers) must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkFreeMemory-memory-00677)
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyBuffer-buffer-00922 ] Object: 0x240000000024 (Type = 9) | Cannot free VkBuffer 0x240000000024[] that is in use by a command buffer. The Vulkan spec states: All submitted commands that refer to buffer, either directly or via a VkBufferView, must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyBuffer-buffer-00922)
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyRenderPass-renderPass-00873 ] Object: 0x60000000006 (Type = 18) | Cannot call vkDestroyRenderPass on VkRenderPass 0x60000000006[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to renderPass must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyRenderPass-renderPass-00873)
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyFramebuffer-framebuffer-00892 ] Object: 0xc000000000c (Type = 24) | Cannot call vkDestroyFramebuffer on VkFramebuffer 0xc000000000c[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to framebuffer must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyFramebuffer-framebuffer-00892)
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyImageView-imageView-01026 ] Object: 0x90000000009 (Type = 14) | Cannot call vkDestroyImageView on VkImageView 0x90000000009[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to imageView must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyImageView-imageView-01026)
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyImage-image-01000 ] Object: 0x70000000007 (Type = 10) | Cannot call vkDestroyImage on VkImage 0x70000000007[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to image, either directly or via a VkImageView, must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyImage-image-01000)
[ERROR] validation layer (Validation):
 [ VUID-vkFreeMemory-memory-00677 ] Object: 0x80000000008 (Type = 8) | Cannot call vkFreeMemory on VkDeviceMemory 0x80000000008[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to memory (via images or buffers) must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkFreeMemory-memory-00677)
[ERROR] validation layer (Validation):
 [ VUID-vkFreeMemory-memory-00677 ] Object: 0xb000000000b (Type = 8) | Cannot call vkFreeMemory on VkDeviceMemory 0xb000000000b[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to memory (via images or buffers) must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkFreeMemory-memory-00677)
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyBuffer-buffer-00922 ] Object: 0xa000000000a (Type = 9) | Cannot free VkBuffer 0xa000000000a[] that is in use by a command buffer. The Vulkan spec states: All submitted commands that refer to buffer, either directly or via a VkBufferView, must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyBuffer-buffer-00922)
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyFence-fence-01120 ] Object: 0x50000000005 (Type = 7) | VkFence 0x50000000005[] is in use. The Vulkan spec states: All queue submission commands that refer to fence must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyFence-fence-01120)
[ERROR] validation layer (Validation):
 [ VUID-vkFreeCommandBuffers-pCommandBuffers-00047 ] Object: 0x555556ba29b0 (Type = 6) | Attempt to free VkCommandBuffer 0x555556ba29b0[] which is in use. The Vulkan spec states: All elements of pCommandBuffers must not be in the pending state (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkFreeCommandBuffers-pCommandBuffers-00047)
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyDescriptorPool-descriptorPool-00303 ] Object: 0x200000000020 (Type = 22) | Cannot call vkDestroyDescriptorPool on VkDescriptorPool 0x200000000020[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to descriptorPool (via any allocated descriptor sets) must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyDescriptorPool-descriptorPool-00303)
[ERROR] validation layer (Validation):
 [ VUID-vkFreeMemory-memory-00677 ] Object: 0xe000000000e (Type = 8) | Cannot call vkFreeMemory on VkDeviceMemory 0xe000000000e[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to memory (via images or buffers) must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkFreeMemory-memory-00677)
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyBuffer-buffer-00922 ] Object: 0xd000000000d (Type = 9) | Cannot free VkBuffer 0xd000000000d[] that is in use by a command buffer. The Vulkan spec states: All submitted commands that refer to buffer, either directly or via a VkBufferView, must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyBuffer-buffer-00922)
[ERROR] validation layer (Validation):
 [ VUID-vkFreeMemory-memory-00677 ] Object: 0x100000000010 (Type = 8) | Cannot call vkFreeMemory on VkDeviceMemory 0x100000000010[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to memory (via images or buffers) must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkFreeMemory-memory-00677)
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyBuffer-buffer-00922 ] Object: 0xf000000000f (Type = 9) | Cannot free VkBuffer 0xf000000000f[] that is in use by a command buffer. The Vulkan spec states: All submitted commands that refer to buffer, either directly or via a VkBufferView, must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyBuffer-buffer-00922)
[ERROR] validation layer (Validation):
 [ VUID-vkFreeMemory-memory-00677 ] Object: 0x120000000012 (Type = 8) | Cannot call vkFreeMemory on VkDeviceMemory 0x120000000012[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to memory (via images or buffers) must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkFreeMemory-memory-00677)
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyBuffer-buffer-00922 ] Object: 0x110000000011 (Type = 9) | Cannot free VkBuffer 0x110000000011[] that is in use by a command buffer. The Vulkan spec states: All submitted commands that refer to buffer, either directly or via a VkBufferView, must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyBuffer-buffer-00922)
[ERROR] validation layer (Validation):
 [ VUID-vkFreeMemory-memory-00677 ] Object: 0x140000000014 (Type = 8) | Cannot call vkFreeMemory on VkDeviceMemory 0x140000000014[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to memory (via images or buffers) must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkFreeMemory-memory-00677)
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyBuffer-buffer-00922 ] Object: 0x130000000013 (Type = 9) | Cannot free VkBuffer 0x130000000013[] that is in use by a command buffer. The Vulkan spec states: All submitted commands that refer to buffer, either directly or via a VkBufferView, must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyBuffer-buffer-00922)
[ERROR] validation layer (Validation):
 [ VUID-vkFreeMemory-memory-00677 ] Object: 0x160000000016 (Type = 8) | Cannot call vkFreeMemory on VkDeviceMemory 0x160000000016[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to memory (via images or buffers) must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkFreeMemory-memory-00677)
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyBuffer-buffer-00922 ] Object: 0x150000000015 (Type = 9) | Cannot free VkBuffer 0x150000000015[] that is in use by a command buffer. The Vulkan spec states: All submitted commands that refer to buffer, either directly or via a VkBufferView, must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyBuffer-buffer-00922)
[ERROR] validation layer (Validation):
 [ VUID-vkFreeMemory-memory-00677 ] Object: 0x180000000018 (Type = 8) | Cannot call vkFreeMemory on VkDeviceMemory 0x180000000018[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to memory (via images or buffers) must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkFreeMemory-memory-00677)
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyBuffer-buffer-00922 ] Object: 0x170000000017 (Type = 9) | Cannot free VkBuffer 0x170000000017[] that is in use by a command buffer. The Vulkan spec states: All submitted commands that refer to buffer, either directly or via a VkBufferView, must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyBuffer-buffer-00922)
[ERROR] validation layer (Validation):
 [ VUID-vkFreeMemory-memory-00677 ] Object: 0x1a000000001a (Type = 8) | Cannot call vkFreeMemory on VkDeviceMemory 0x1a000000001a[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to memory (via images or buffers) must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkFreeMemory-memory-00677)
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyBuffer-buffer-00922 ] Object: 0x190000000019 (Type = 9) | Cannot free VkBuffer 0x190000000019[] that is in use by a command buffer. The Vulkan spec states: All submitted commands that refer to buffer, either directly or via a VkBufferView, must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyBuffer-buffer-00922)
[ERROR] validation layer (Validation):
 [ VUID-vkFreeMemory-memory-00677 ] Object: 0x1c000000001c (Type = 8) | Cannot call vkFreeMemory on VkDeviceMemory 0x1c000000001c[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to memory (via images or buffers) must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkFreeMemory-memory-00677)
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyBuffer-buffer-00922 ] Object: 0x1b000000001b (Type = 9) | Cannot free VkBuffer 0x1b000000001b[] that is in use by a command buffer. The Vulkan spec states: All submitted commands that refer to buffer, either directly or via a VkBufferView, must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyBuffer-buffer-00922)
[ERROR] validation layer (Validation):
 [ VUID-vkFreeMemory-memory-00677 ] Object: 0x1e000000001e (Type = 8) | Cannot call vkFreeMemory on VkDeviceMemory 0x1e000000001e[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to memory (via images or buffers) must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkFreeMemory-memory-00677)
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyBuffer-buffer-00922 ] Object: 0x1d000000001d (Type = 9) | Cannot free VkBuffer 0x1d000000001d[] that is in use by a command buffer. The Vulkan spec states: All submitted commands that refer to buffer, either directly or via a VkBufferView, must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyBuffer-buffer-00922)
tests/cases/matrices_uniform_draw.amber: Vulkan::Calling vkWaitForFences Timeout

Summary of Failures:
  tests/cases/matrices_uniform_draw.amber

Summary: 0 pass, 1 fail
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyDevice-device-00378 ] Object: 0x220000000022 (Type = 17) | OBJ ERROR : For VkDevice 0x5555567485c8[], VkPipelineLayout 0x220000000022[] has not been destroyed. The Vulkan spec states: All child objects created on device must have been destroyed prior to destroying device (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyDevice-device-00378)
[ERROR] validation layer (Validation):
 [ VUID-vkDestroyDevice-device-00378 ] Object: 0x230000000023 (Type = 19) | OBJ ERROR : For VkDevice 0x5555567485c8[], VkPipeline 0x230000000023[] has not been destroyed. The Vulkan spec states: All child objects created on device must have been destroyed prior to destroying device (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyDevice-device-00378)
../../third_party/swiftshader/src/Vulkan/VkDeviceMemory.cpp:196 ABORT: ASSERT(buffer)

```

```
[Switching to Thread 0x7fffaf7fe700 (LWP 248379)]
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff6f098ad in __GI_abort () at abort.c:79
#2  0x00007ffff3848952 in vk::abort (format=0x7ffff53fe3b3 "%s:%d ABORT: ASSERT(%s)\n\n") at ../../third_party/swiftshader/src/Vulkan/VkDebug.cpp:139
#3  0x00007ffff3852fd4 in vk::DeviceMemory::getOffsetPointer (this=0x555556ba8368, pOffset=0) at ../../third_party/swiftshader/src/Vulkan/VkDeviceMemory.cpp:196
#4  0x00007ffff385f5c9 in vk::Image::getTexelPointer (this=0x555556ba84b8, offset=..., subresource=...) at ../../third_party/swiftshader/src/Vulkan/VkImage.cpp:473
#5  0x00007ffff3861b29 in vk::ImageView::getOffsetPointer (this=0x555556b47828, offset=..., aspect=VK_IMAGE_ASPECT_COLOR_BIT, mipLevel=0, layer=0, usage=vk::ImageView::RAW)
    at ../../third_party/swiftshader/src/Vulkan/VkImageView.cpp:293
#6  0x00007ffff374407b in sw::Renderer::draw (this=0x7fffa4001360, context=0x7fffaf7fd210, indexType=2944392528, count=2, baseVertex=0, events=0x555556ba4658, instanceID=0, viewID=0, indexBuffer=0x0, framebufferExtent=..., 
    pushConstants=..., update=true) at ../../third_party/swiftshader/src/Device/Renderer.cpp:361
#7  0x00007ffff3832860 in vk::DrawBase::draw (this=0x555556c9eca0, executionState=..., indexed=false, count=4, instanceCount=1, first=0, vertexOffset=0, firstInstance=0)
    at ../../third_party/swiftshader/src/Vulkan/VkCommandBuffer.cpp:617
#8  0x00007ffff38329f4 in vk::Draw::play (this=0x555556c9eca0, executionState=...) at ../../third_party/swiftshader/src/Vulkan/VkCommandBuffer.cpp:638
#9  0x00007ffff3831117 in vk::CommandBuffer::submit (this=0x555556ba29b8, executionState=...) at ../../third_party/swiftshader/src/Vulkan/VkCommandBuffer.cpp:1619
#10 0x00007ffff3871ee8 in vk::Queue::submitQueue (this=0x55555667aa50, task=...) at ../../third_party/swiftshader/src/Vulkan/VkQueue.cpp:139
#11 0x00007ffff38720a9 in vk::Queue::taskLoop (this=0x55555667aa50, scheduler=0x555556a78a60) at ../../third_party/swiftshader/src/Vulkan/VkQueue.cpp:179
#12 0x00007ffff38738d9 in std::__invoke_impl<void, void (vk::Queue::*)(marl::Scheduler*), vk::Queue*, marl::Scheduler*> (__f=
    @0x555556a7a818: (void (vk::Queue::*)(class vk::Queue * const, class marl::Scheduler *)) 0x7ffff3871fd6 <vk::Queue::taskLoop(marl::Scheduler*)>, __t=@0x555556a7a810: 0x55555667aa50, __args#0=@0x555556a7a808: 0x555556a78a60)
    at /usr/include/c++/8/bits/invoke.h:73
```

--- a/src/engine.h
+++ b/src/engine.h
@@ -32,7 +32,7 @@ namespace amber {
 /// EngineData stores information used during engine execution.
 struct EngineData {
   /// The timeout to use for fences, in milliseconds.
-  uint32_t fence_timeout_ms = 1000;
+  uint32_t fence_timeout_ms = 10000;
 };
 
 /// Abstract class which describes a backing engine for Amber.

--- a/src/vulkan/command_buffer.cc
+++ b/src/vulkan/command_buffer.cc
@@ -76,8 +76,6 @@ Result CommandBuffer::SubmitAndReset(uint32_t timeout_ms) {
   if (device_->GetPtrs()->vkEndCommandBuffer(command_) != VK_SUCCESS)
     return Result("Vulkan::Calling vkEndCommandBuffer Fail");
 
-  guarded_ = false;
-
   if (device_->GetPtrs()->vkResetFences(device_->GetVkDevice(), 1, &fence_) !=
       VK_SUCCESS) {
     return Result("Vulkan::Calling vkResetFences Fail");
@@ -103,15 +101,17 @@ Result CommandBuffer::SubmitAndReset(uint32_t timeout_ms) {
   if (device_->GetPtrs()->vkResetCommandBuffer(command_, 0) != VK_SUCCESS)
     return Result("Vulkan::Calling vkResetCommandBuffer Fail");
 
+  guarded_ = false;
+
   return {};
 }
 
 void CommandBuffer::Reset() {
   if (guarded_) {
     device_->GetPtrs()->vkEndCommandBuffer(command_);
+    device_->GetPtrs()->vkResetCommandBuffer(command_, 0);
     guarded_ = false;
   }
-  device_->GetPtrs()->vkResetCommandBuffer(command_, 0);
 }
 
 CommandBufferGuard::CommandBufferGuard(CommandBuffer* buffer)

--- a/src/vulkan/command_buffer.cc
+++ b/src/vulkan/command_buffer.cc
@@ -60,8 +60,6 @@ Result CommandBuffer::Initialize() {
 }
 
 Result CommandBuffer::BeginRecording() {
-  guarded_ = true;
-
   VkCommandBufferBeginInfo command_begin_info = VkCommandBufferBeginInfo();
   command_begin_info.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO;
   command_begin_info.flags = VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT;
@@ -69,6 +67,7 @@ Result CommandBuffer::BeginRecording() {
       VK_SUCCESS) {
     return Result("Vulkan::Calling vkBeginCommandBuffer Fail");
   }
+  guarded_ = true;
 
   return {};
 }
@@ -76,6 +75,8 @@ Result CommandBuffer::BeginRecording() {
 Result CommandBuffer::SubmitAndReset(uint32_t timeout_ms) {
   if (device_->GetPtrs()->vkEndCommandBuffer(command_) != VK_SUCCESS)
     return Result("Vulkan::Calling vkEndCommandBuffer Fail");
+
+  guarded_ = false;
 
   if (device_->GetPtrs()->vkResetFences(device_->GetVkDevice(), 1, &fence_) !=
       VK_SUCCESS) {
@@ -102,7 +103,6 @@ Result CommandBuffer::SubmitAndReset(uint32_t timeout_ms) {
   if (device_->GetPtrs()->vkResetCommandBuffer(command_, 0) != VK_SUCCESS)
     return Result("Vulkan::Calling vkResetCommandBuffer Fail");
 
-  guarded_ = false;
 
   return {};
 }
@@ -110,9 +110,9 @@ Result CommandBuffer::SubmitAndReset(uint32_t timeout_ms) {
 void CommandBuffer::Reset() {
   if (guarded_) {
     device_->GetPtrs()->vkEndCommandBuffer(command_);
-    device_->GetPtrs()->vkResetCommandBuffer(command_, 0);
     guarded_ = false;
   }
+  device_->GetPtrs()->vkResetCommandBuffer(command_, 0);
 }
 
 CommandBufferGuard::CommandBufferGuard(CommandBuffer* buffer)

--- a/src/vulkan/command_buffer.cc
+++ b/src/vulkan/command_buffer.cc
@@ -90,6 +90,8 @@ Result CommandBuffer::SubmitAndReset(uint32_t timeout_ms) {
     return Result("Vulkan::Calling vkQueueSubmit Fail");
   }
 
+  guarded_ = false;
+
   VkResult r = device_->GetPtrs()->vkWaitForFences(
       device_->GetVkDevice(), 1, &fence_, VK_TRUE,
       static_cast<uint64_t>(timeout_ms) * 1000ULL * 1000ULL /* nanosecond */);
@@ -100,8 +102,6 @@ Result CommandBuffer::SubmitAndReset(uint32_t timeout_ms) {
 
   if (device_->GetPtrs()->vkResetCommandBuffer(command_, 0) != VK_SUCCESS)
     return Result("Vulkan::Calling vkResetCommandBuffer Fail");
-
-  guarded_ = false;
 
   return {};
 }
@@ -117,19 +117,17 @@ void CommandBuffer::Reset() {
 CommandBufferGuard::CommandBufferGuard(CommandBuffer* buffer)
     : buffer_(buffer) {
   assert(!buffer_->guarded_);
-
   result_ = buffer_->BeginRecording();
 }
 
 CommandBufferGuard::~CommandBufferGuard() {
-  buffer_->Reset();
+  if (buffer_->guarded_)
+    buffer_->Reset();
 }
 
 Result CommandBufferGuard::Submit(uint32_t timeout_ms) {
   assert(buffer_->guarded_);
-
-  result_ = buffer_->SubmitAndReset(timeout_ms);
-  return result_;
+  return buffer_->SubmitAndReset(timeout_ms);
 }
 
 }  // namespace vulkan

--- a/src/vulkan/command_buffer.cc
+++ b/src/vulkan/command_buffer.cc
@@ -103,7 +103,6 @@ Result CommandBuffer::SubmitAndReset(uint32_t timeout_ms) {
   if (device_->GetPtrs()->vkResetCommandBuffer(command_, 0) != VK_SUCCESS)
     return Result("Vulkan::Calling vkResetCommandBuffer Fail");
 
-
   return {};
 }
 

--- a/src/vulkan/command_buffer.cc
+++ b/src/vulkan/command_buffer.cc
@@ -26,6 +26,8 @@ CommandBuffer::CommandBuffer(Device* device, CommandPool* pool)
     : device_(device), pool_(pool) {}
 
 CommandBuffer::~CommandBuffer() {
+  Reset();
+
   if (fence_ != VK_NULL_HANDLE)
     device_->GetPtrs()->vkDestroyFence(device_->GetVkDevice(), fence_, nullptr);
 
@@ -58,6 +60,8 @@ Result CommandBuffer::Initialize() {
 }
 
 Result CommandBuffer::BeginRecording() {
+  guarded_ = true;
+
   VkCommandBufferBeginInfo command_begin_info = VkCommandBufferBeginInfo();
   command_begin_info.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO;
   command_begin_info.flags = VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT;
@@ -98,34 +102,34 @@ Result CommandBuffer::SubmitAndReset(uint32_t timeout_ms) {
   if (device_->GetPtrs()->vkResetCommandBuffer(command_, 0) != VK_SUCCESS)
     return Result("Vulkan::Calling vkResetCommandBuffer Fail");
 
+  guarded_ = false;
+
   return {};
 }
 
 void CommandBuffer::Reset() {
-  device_->GetPtrs()->vkEndCommandBuffer(command_);
-  device_->GetPtrs()->vkResetCommandBuffer(command_, 0);
+  if (guarded_) {
+    device_->GetPtrs()->vkEndCommandBuffer(command_);
+    device_->GetPtrs()->vkResetCommandBuffer(command_, 0);
+    guarded_ = false;
+  }
 }
 
 CommandBufferGuard::CommandBufferGuard(CommandBuffer* buffer)
     : buffer_(buffer) {
   assert(!buffer_->guarded_);
 
-  buffer_->guarded_ = true;
   result_ = buffer_->BeginRecording();
 }
 
 CommandBufferGuard::~CommandBufferGuard() {
-  if (buffer_->guarded_) {
-    buffer_->Reset();
-    buffer_->guarded_ = false;
-  }
+  buffer_->Reset();
 }
 
 Result CommandBufferGuard::Submit(uint32_t timeout_ms) {
   assert(buffer_->guarded_);
 
   result_ = buffer_->SubmitAndReset(timeout_ms);
-  buffer_->guarded_ = false;
   return result_;
 }
 

--- a/src/vulkan/pipeline.h
+++ b/src/vulkan/pipeline.h
@@ -120,7 +120,7 @@ class Pipeline {
   std::vector<DescriptorSetInfo> descriptor_set_info_;
   std::vector<VkPipelineShaderStageCreateInfo> shader_stage_info_;
 
-  uint32_t fence_timeout_ms_ = 100;
+  uint32_t fence_timeout_ms_ = 1000;
   bool descriptor_related_objects_already_created_ = false;
   std::unordered_map<VkShaderStageFlagBits,
                      std::string,

--- a/tests/run_tests.py
+++ b/tests/run_tests.py
@@ -78,11 +78,6 @@ SUPPRESSIONS_SWIFTSHADER = [
   "draw_triangle_list_in_r8g8b8a8_snorm_color_frame.vkscript",
   # No supporting device for Float16Int8Features
   "float16.amber",
-  # SEGV: github.com/google/amber/issues/726
-  "matrices_uniform_draw.amber",
-  # SEGV: github.com/google/amber/issues/725
-  "multiple_ssbo_update_with_graphics_pipeline.vkscript",
-  "multiple_ssbo_with_sparse_descriptor_set_in_compute_pipeline_less_than_4.vkscript",
   # Exceeded maxBoundDescriptorSets limit of physical device
   "multiple_ssbo_with_sparse_descriptor_set_in_compute_pipeline.vkscript",
   # shaderStorageImageWriteWithoutFormat but is not enabled on the device

--- a/tests/run_tests.py
+++ b/tests/run_tests.py
@@ -78,6 +78,9 @@ SUPPRESSIONS_SWIFTSHADER = [
   "draw_triangle_list_in_r8g8b8a8_snorm_color_frame.vkscript",
   # No supporting device for Float16Int8Features
   "float16.amber",
+  # SEGV: github.com/google/amber/issues/725
+  "multiple_ssbo_update_with_graphics_pipeline.vkscript",
+  "multiple_ssbo_with_sparse_descriptor_set_in_compute_pipeline_less_than_4.vkscript",
   # Exceeded maxBoundDescriptorSets limit of physical device
   "multiple_ssbo_with_sparse_descriptor_set_in_compute_pipeline.vkscript",
   # shaderStorageImageWriteWithoutFormat but is not enabled on the device

